### PR TITLE
Exo API

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
# Exo api
- `Exo api` is like go cs but in exo with cobra instead of urfave
```
Exoscale api

Usage:
  exo api [command]

Available Commands:
  account
  address
  apis
  async-job
  event
  network
  nic
  offerings
  security-group
  snapshot
  ssh
  tags
  template
  user
  vm
  vm-group
  volume
  zone

Flags:
  -d, --debug   debug mode on
  -h, --help    help for api

Global Flags:
  -A, --account string   Account to use in config file [env EXOSCALE_ACCOUNT]
  -C, --config string    Specify an alternate config file [env EXOSCALE_CONFIG]

Use "exo api [command] --help" for more information about a command.
```

### An example:
```
$ exo api zone list
{
  "count": 4,
  "zone": [
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "1128bd56-b4d9-4ac6-a7b9-c715b187ce11",
      "localstorageenabled": true,
      "name": "ch-gva-2",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "ccb0a60c-79c8-3230-ab8b-8bdbe8c45bb7"
    },
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "91e5e9e4-c9ed-4b76-bee4-427004b3baf9",
      "localstorageenabled": true,
      "name": "ch-dk-2",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "fe63f9cb-ff75-31d3-8c46-3631f7fcd533"
    },
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "4da1b188-dcd6-4ff5-b7fd-bde984055548",
      "localstorageenabled": true,
      "name": "at-vie-1",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "26d84c22-f66d-377e-93ab-987ef477cab3"
    },
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "35eb7739-d19e-45f7-a581-4687c54d6d02",
      "localstorageenabled": true,
      "name": "de-fra-1",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "67885c8d-7230-3bac-81ed-e5f4f286f55b"
    }
  ]
}
```

### CODE REVIEW

**Comment by pierre-emmanuelJ** [comment]
@mcorbin Good news for you ! Working well

File: cmd/exo/cmd/api_layout.go [hunk 1]
```diff
@@ -7,142 +7,143 @@ import (
...
+		{&egoscale.RegisterUserKeys{}, false, ""},
+	},
+	"security-group,sg": {
+		{&egoscale.AuthorizeSecurityGroupEgress{}, false, "authorizeEgress"},
+		{&egoscale.AuthorizeSecurityGroupIngress{}, false, "authorizeIngress"},
+		{&egoscale.CreateSecurityGroup{}, false, "create"},
+		{&egoscale.DeleteSecurityGroup{}, false, "delete"},
+		{&egoscale.ListSecurityGroups{}, false, "list"},
+		{&egoscale.RevokeSecurityGroupEgress{}, false, "RevokeEgress"},
+		{&egoscale.RevokeSecurityGroupIngress{}, false, "RevokeIngress"},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: lower _r_

File: cmd/exo/cmd/api_layout.go [hunk 2]
```diff
@@ -7,142 +7,143 @@ import (
...
-		{&egoscale.CreateUser{}, true},
-		{&egoscale.DeleteUser{}, true},
+		{&egoscale.CreateUser{}, true, ""},
+		{&egoscale.DeleteUser{}, true, ""},
 		//{&egoscale.DisableUser{}, true},
 		//{&egoscale.GetUser{}, true},
-		{&egoscale.UpdateUser{}, true},
-		{&egoscale.ListUsers{}, false},
-		{&egoscale.RegisterUserKeys{}, false},
-	},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: I'd remove the _hidden_ ones altogether

**Comment by greut** [comment]
Does it generate the doc?

File: cmd/exo/cmd/api_layout.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/api_layout.go]: ohww thanks

**Comment by pierre-emmanuelJ** [comment]
I'll try to generate doc

**Comment by greut** [comment]
@ste-m 

```console
% exo api zone list --keyword=ch
```
```json
{
  "count": 2,
  "zone": [
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "1747ef5e-5451-41fd-9f1a-58913bae9702",
      "localstorageenabled": true,
      "name": "ch-gva-2",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "f9a2983b-42e5-3b12-ae74-0b1f54cd6204"
    },
    {
      "allocationstate": "Enabled",
      "dhcpprovider": "VirtualRouter",
      "id": "381d0a95-ed4a-4ad9-b41c-b97073c1a433",
      "localstorageenabled": true,
      "name": "ch-dk-2",
      "networktype": "Basic",
      "securitygroupsenabled": true,
      "zonetoken": "23a24359-121a-38af-a938-e225c97c397b"
    }
  ]
}
```

File: cmd/exo/cmd/api.go [hunk 3]
```diff
@@ -271,3 +273,45 @@ func buildFlags(method egoscale.Command, cmd *cobra.Command) {
...
 }
+
+func printResponseHelp(out io.Writer, response interface{}) {
+	value := reflect.ValueOf(response)
+	typeof := reflect.TypeOf(response)
+
+	w := tabwriter.NewWriter(out, 0, 0, 1, ' ', tabwriter.FilterHTML)
+	if _, err := fmt.Fprintln(w, "FIELD\tTYPE\tDOCUMENTATION"); err != nil {
+		log.Fatal(err)
+	}
```
Comment (greut) [inline cmd/exo/cmd/api.go]: #296

File: cmd/exo/cmd/api_layout.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	},
+	"address": {
+		{&egoscale.AssociateIPAddress{}, "associate"},
+		{&egoscale.DisassociateIPAddress{}, "disassociate"},
+		{&egoscale.ListPublicIPAddresses{}, "list"},
+		{&egoscale.UpdateIPAddress{}, "update"},
+	},
+	"async-job,aj": {
+		{&egoscale.QueryAsyncJobResult{}, ""},
+		{&egoscale.ListAsyncJobs{}, ""},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: list query ?

File: cmd/exo/cmd/api_layout.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		{&egoscale.ListSSHKeyPairs{}, "list"},
+		{&egoscale.CreateSSHKeyPair{}, "create"},
+		{&egoscale.DeleteSSHKeyPair{}, "delete"},
+		{&egoscale.ResetSSHKeyForVirtualMachine{}, "reset"},
+	},
+	"affinity-group,ag": {
+		{&egoscale.CreateAffinityGroup{}, "create"},
+		{&egoscale.DeleteAffinityGroup{}, "delete"},
+		{&egoscale.ListAffinityGroups{}, "list"},
+		{&egoscale.UpdateVMAffinityGroup{}, ""},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: move this into the `vm` group

File: cmd/exo/cmd/api_layout.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		{&egoscale.ListHosts{}, ""},
+		{&egoscale.UpdateHost{}, ""},
+	},
+	"reversedns": {
+		{&egoscale.DeleteReverseDNSFromPublicIPAddress{}, "deleteFromAddress"},
+		{&egoscale.QueryReverseDNSForPublicIPAddress{}, "queryForAddress"},
+		{&egoscale.UpdateReverseDNSForPublicIPAddress{}, "updateForAddress"},
+		{&egoscale.DeleteReverseDNSFromVirtualMachine{}, "deleteFromVM"},
+		{&egoscale.QueryReverseDNSForVirtualMachine{}, "queryForVM"},
+		{&egoscale.UpdateReverseDNSForVirtualMachine{}, "updateForVM"},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: move those into the `vm ` and the `ip` group

File: cmd/exo/cmd/api_layout.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+		{&egoscale.ListEvents{}, "list"},
+	},
+	"offerings": {
+		{&egoscale.ListResourceDetails{}, "listDetails"},
+		{&egoscale.ListResourceLimits{}, "listLimits"},
+		{&egoscale.ListServiceOfferings{}, "list"},
+	},
+	"host": {
+		{&egoscale.ListHosts{}, ""},
+		{&egoscale.UpdateHost{}, ""},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: we don't need those

File: cmd/exo/cmd/api_layout.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+		{&egoscale.CopyTemplate{}, ""},
+		{&egoscale.CreateTemplate{}, ""},
+		{&egoscale.ListTemplates{}, "list"},
+		{&egoscale.PrepareTemplate{}, ""},
+		{&egoscale.RegisterTemplate{}, ""},
+		{&egoscale.ListOSCategories{}, ""},
+	},
+	"account": {
+		{&egoscale.EnableAccount{}, ""},
+		{&egoscale.DisableAccount{}, ""},
```
Comment (greut) [inline cmd/exo/cmd/api_layout.go]: remove

**Comment by greut** [comment]
screenshot from 2018-08-17 15-35-12

**Comment by pierre-emmanuelJ** [comment]
localhost_1313_egoscale_cli_exo_api_vm_ ipad pro
Working for Sub Command but I have tu put in layout for group command

**Comment by pierre-emmanuelJ** [comment]
image

**Comment by greut** [comment]
nitpick®

- plural vs singular forms: `apis`, `offerings`, `tags` vs the others.
- I'd love to have something like `exo api listZones` in addition  to `exo api zone list`

**Comment by pierre-emmanuelJ** [comment]
`exo api zone list` is `exo api zone ls` but you speak about to move in sigle command all category with one command ?
`apis` => `api`, `offerings` => `offering`, ...

**Pull request merged**